### ISSUE TITLE
How to access gSampleRate from within a library?

### ISSUE THREAD

**Issue by antoniograzioli** [VCVRack/Rack#259]
I am developing some oscillators using some external libraries and samples
How can I access the global gSampleRate from within one of these libararies?

**Comment by antoniotuzzi** [comment]
you need at least a call to a "setSampleRate" of some kind, somewhere in the lib

or if the lib is compiled within your code, just use the gSampleRate

or declare an extern float gSampleRate in the lib, compile the lib, and hope the symbol is recognised at linking time with your plugins and rack libs

**Comment by AndrewBelt** [comment]
Thanks for reminding me. I was going to remove `gSampleRate` in favor of a getter function. It essentially has no overhead (although it does have a nonzero overhead, so don't call it 1000 times in `step()`).
Added `engineGetSampleRate()` with 9dbadfb

**Comment by antoniotuzzi** [comment]
@AndrewBelt a plugin should access only during instantiation to the

` engineGetSampleRate();`

at this time if plugin needs the value, should store as member and use it


if the engine changes the SR the event 

`onSampleRateChange()`

Is all we need; is the place where plugins will update the stored value and do recalculations

**Comment by AndrewBelt** [comment]
It's not a significant problem to call this every `step()`. It might even be faster for cache reasons. So don't worry of your plugin just directly calls it

**Comment by antoniograzioli** [comment]
I see you already swithced from **gSampleRate** to **engineGetSampleRate()** in sources
all of  sudden Fundamentals don't compile and work anymore, of course
all plugins referencing gSampeRate need to replace it with engineGetSampleRate()

**Comment by AndrewBelt** [comment]
I'll get around to that in the next couple hours

**Comment by antoniograzioli** [comment]
yeah, but all plugin devs will have to change it as well (I already did) ;-)

**Comment by mcclure** [comment]
Minor suggestion: if you are going to make backward-incompatible API changes sometimes, maybe recommend in the plugin tutorial README that users "checkout" the most recent release tag before building the plugin. Alternately, maybe consider creating a "release" branch which always points to the most recent released tag.

I checked out the tutorial and rack (as the tutorial build steps recommend) just now, but it did not compile, because it was expecting engineGetSampleRate(). However I did not really want to be compiling against git HEAD, I wanted to be compiling against 0.4.0 because that is what I have installed on my computer, as soon as I checked out tag v0.4.0 it worked great…
